feature_engine/timeseries/forecasting/lag_features.py

```diff
@@ -0,0 +1,292 @@
+                            f"as one of the elements in periods instead."
+                        )
+
+        if missing_values not in ["raise", "ignore"]:
+            raise ValueError(
+                "missing_values takes only values 'raise' or 'ignore'."
+                f"Got {missing_values} instead."
+            )
+
+        if not isinstance(drop_original, bool):
+            raise ValueError(
+                "drop_original takes only boolean values True and False. "
+                f"Got {drop_original} instead."
+            )
+
+        self.variables = _check_input_parameter_variables(variables)
+        self.periods = periods
+        self.freq = freq
+        self.missing_values = missing_values
+        self.drop_original = drop_original
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        This transformer does not learn parameters.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The training dataset.
+
+        y: pandas Series, default=None
+            y is not needed in this imputation. You can pass None or y.
+        """
+        # check input dataframe
+        X = _is_dataframe(X)
+
+        # find or check for numerical variables
+        self.variables_ = _find_or_check_numerical_variables(X, self.variables)
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        self.n_features_in_ = X.shape[1]
+
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Adds lag features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: Pandas dataframe, shape = [n_samples, n_features + n_operations]
+            The dataframe with the original variables plus the new variables.
+        """
+        # Check method fit has been called
+        check_is_fitted(self)
+
+        # check if 'X' is a dataframe
+        X = _is_dataframe(X)
+
+        # Check if input data contains same number of columns as dataframe used to fit.
+        _check_input_matches_training_df(X, self.n_features_in_)
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        if isinstance(self.freq, list):
+            df_ls = []
+            for fr in self.freq:
+                tmp = X[self.variables_].shift(
+                    freq=fr,
+                    axis=0,
+                )
+                df_ls.append(tmp)
+            tmp = pd.concat(df_ls, axis=1)
+
+        elif isinstance(self.periods, list):
```

Comment:
The user is still able to pass a list for both `periods` and `freq`.  With this implementation the list passed to `periods` is ignored and the list passed to `freq` takes precedence. 

`LagFeatures(periods=[1, 2, 3, 4], freq=['15T','30T']).fit(df).transform(df)`

Still returns a valid output despite the fact that incompatible inputs have been provided. 

There needs to be an agreement as to what the behaviour should be if a list is passed to both `periods` and `freq`. The most reasonable options imo would be:

1. A cartesian product of the two options (so you generate `len(periods)*len(freq)` new lagged features)
2. You require that the length of the lists for both parameters are the same and mimic the same behaviour as for when you do not provide a list of arguments (i.e., you map the first element in `periods` and you create a lag by the amount specified in the first element in `freq` and so on). 
3. Raise an error and don't allow a list to be passed for both the `freq` and `periods` argument.

@solegalli - what do you think? I think option 2 or 3 is probably the more intuitive and expected behaviour.

Reply:
@KishManani good point. 

I think the current implementation does not do both. Because it says **if** freq is list, do this, **elif** period is list do that.

I was mostly inclined to give the param `freq` priority over `periods`. So, if you pass a list to freq, we ignore periods (and mention that in the docs). Mostly because I do not see a case where you would want both. If you pass the frequency, you kind of know very well what you want, so why would you pass a list of freq and a list of periods?

But then, I haven't lagged a lot of features myself (lol), so if you can see a case where a user would want to pass both, happy to change.

For simplicity, we could also go with 3, because this class would already simplify the user's life a lot, so if they want features with a list of freq and a list of periods, they may as well call the transformer twice, once for each.

@KishManani thoughts?

Reply:
@solegalli thanks for highlighting the comment in the docstring. I missed that. It says the following in the docstring `If freq is not None, then this parameter overrides the parameter periods`. I don't think this is true. The current implementation only overrides `periods` when `freq` is a `list`. For example the following:

`LagFeatures(periods=3, freq='2D').fit(df).transform(df)`

would shift the index by 6 days (which is also the expected behaviour and the same as what you get with `.shift()`. This is an example where `freq` is not `None` and did not override the `periods` parameter. 

Given the above example there is a way in which it could be sensible to have a `list` for both `periods` and `freq`. But in my opinion this is redundant because we could have just provided in the example above `freq='6D'`. 

Likewise the following example could generate confusion:

`LagFeatures(periods=[1,2,3], freq='D').fit(df).transform(df)`

We might expect this to lag by 1 day, 2 days, and 3 days. But the `freq` parameter would be completely ignored in the current implementation in this example.

A simpler solution might be to restrict the user to only allow them to provide inputs for `periods` or `freq` but not both and raise an error if this is tried. If the user wanted to lag by 3 days they could just use `freq='3D'` rather than `periods=3` and `freq='D'`.

Reply:
Enlightening. I did not know about this. That's why we need the expert :) Thank you @KishManani 

I agree and 'll fix accordingly.

Reply:
I implemented the check that raises an error if both `periods` and `freq` have values. I also built a test for the corresponding check.

I changed the `periods` default value to `None`. It was `1`. Otherwise, an error is raised whenever the user solely inputs a value for `freq`. It seems that this would be a common practice.